core/src/com/google/inject/internal/InternalFactory.java

```diff
@@ -21,14 +21,21 @@
 import com.google.inject.Provider;
 import com.google.inject.spi.Dependency;
 import java.lang.invoke.MethodHandle;
-import java.lang.invoke.MethodHandles;
 
 /**
  * Creates objects which will be injected.
  *
  * @author crazybob@google.com (Bob Lee)
  */
-interface InternalFactory<T> {
+abstract class InternalFactory<T> {
+
+  /**
+   * This will be a {@link MethodHandle} if the handle is always cachable, or a {@code
+   * MethodHandle[2]} if the handle is conditionally cachable based on the linked status. In which
+   * case index 0 is the handle for when not linked and index 1 is for when it is linked.
+   * 
+   */
+  private volatile Object handleCacheResult;
```

Comment:
Would it be wildly inefficient to make this more type-safe, some kind of MethodHandleRef interface type w/ implementations: SimpleRef and LinkedRef, where SimpleRef is just a passthru & LinkedRef has two volatile fields of linkedRef & unlinkedRef?

(This would also negate the question below about cross-thread safety of setting items in the array.)

Reply:
Done.  I was concerned about adding memory overhead to the case where methodhandles were not in in use but i believe i found a solution to that.  This should resolve your (legitimate) threadsafety concerns as well.

Basically we can create a small type hierarchy that can handle the different transitions and then a simple compare-and-swap to pick a winner.  The cost is allocating some tiny objects but since we onlyt do it when caching, i don't really think it will matter.